// nextra/dist/client/components/callout.js
import { jsx, jsxs } from "react/jsx-runtime";
import { c as _c } from "react-compiler-runtime";
import cn from "clsx";
import { GitHubCautionIcon, GitHubImportantIcon, GitHubNoteIcon, GitHubTipIcon, GitHubWarningIcon } from "../icons/index.js";
const TypeToEmoji = {
  default: /* @__PURE__ */ jsx(GitHubTipIcon, { height: ".8em", className: "x:mt-[.3em]" }),
  error: /* @__PURE__ */ jsx(GitHubCautionIcon, { height: ".8em", className: "x:mt-[.3em]" }),
  info: /* @__PURE__ */ jsx(GitHubNoteIcon, { height: ".8em", className: "x:mt-[.3em]" }),
  warning: /* @__PURE__ */ jsx(GitHubWarningIcon, { height: ".8em", className: "x:mt-[.3em]" }),
  important: /* @__PURE__ */ jsx(GitHubImportantIcon, { height: ".8em", className: "x:mt-[.3em]" })
};
const classes = {
  default: cn("x:bg-green-100 x:dark:bg-green-900/30", "x:text-green-700 x:dark:text-green-500", "x:border-green-700 x:dark:border-green-800"),
  error: cn("x:bg-red-100 x:dark:bg-red-900/30", "x:text-red-700 x:dark:text-red-500", "x:border-red-700 x:dark:border-red-600"),
  info: cn("x:bg-blue-100 x:dark:bg-blue-900/30", "x:text-blue-700 x:dark:text-blue-400", "x:border-blue-700 x:dark:border-blue-600"),
  warning: cn("x:bg-yellow-50 x:dark:bg-yellow-700/30", "x:text-yellow-700 x:dark:text-yellow-500", "x:border-yellow-700"),
  important: cn("x:bg-purple-100 x:dark:bg-purple-900/30", "x:text-purple-600 x:dark:text-purple-400", "x:border-purple-600")
};
const Callout = (t0) => {
  const $ = _c(18);
  let className;
  let props;
  let t1;
  let t2;
  if ($[0] !== t0) {
    ({
      className,
      type: t1,
      emoji: t2,
      ...props
    } = t0);
    $[0] = t0;
    $[1] = className;
    $[2] = props;
    $[3] = t1;
    $[4] = t2;
  } else {
    className = $[1];
    props = $[2];
    t1 = $[3];
    t2 = $[4];
  }
  const type = t1 === void 0 ? "default" : t1;
  const emoji = t2 === void 0 ? type && TypeToEmoji[type] : t2;
  const t3 = type && classes[type];
  let t4;
  if ($[5] !== t3) {
    t4 = cn("nextra-callout x:overflow-x-auto x:not-first:mt-[1.25em] x:flex x:rounded-lg x:border x:py-[.5em] x:pe-[1em]", "x:contrast-more:border-current!", t3);
    $[5] = t3;
    $[6] = t4;
  } else {
    t4 = $[6];
  }
  let t5;
  if ($[7] !== emoji) {
    t5 = /* @__PURE__ */ jsx("div", { className: "x:select-none x:text-[1.25em] x:ps-[.6em] x:pe-[.4em]", style, "data-pagefind-ignore": "all", children: emoji });
    $[7] = emoji;
    $[8] = t5;
  } else {
    t5 = $[8];
  }
  let t6;
  if ($[9] !== className) {
    t6 = cn("x:w-full x:min-w-0", className);
    $[9] = className;
    $[10] = t6;
  } else {
    t6 = $[10];
  }
  let t7;
  if ($[11] !== props || $[12] !== t6) {
    t7 = /* @__PURE__ */ jsx("div", { className: t6, ...props });
    $[11] = props;
    $[12] = t6;
    $[13] = t7;
  } else {
    t7 = $[13];
  }
  let t8;
  if ($[14] !== t4 || $[15] !== t5 || $[16] !== t7) {
    t8 = /* @__PURE__ */ jsxs("div", { className: t4, children: [
      t5,
      t7
    ] });
    $[14] = t4;
    $[15] = t5;
    $[16] = t7;
    $[17] = t8;
  } else {
    t8 = $[17];
  }
  return t8;
};
const style = {
  fontFamily: '"Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"'
};
export {
  Callout
};
