// nextra/dist/client/components/select.js
"use client";
import { jsx, jsxs } from "react/jsx-runtime";
import { c as _c } from "react-compiler-runtime";
import { Listbox, ListboxButton, ListboxOption, ListboxOptions } from "@headlessui/react";
import cn from "clsx";
import { Fragment } from "react";
import { CheckIcon } from "../icons/index.js";
const Select = (t0) => {
  const $ = _c(18);
  const {
    options,
    onChange,
    selectedOption,
    value,
    title,
    className,
    anchor: t1
  } = t0;
  let t2;
  if ($[0] !== t1) {
    t2 = t1 === void 0 ? {
      to: "top start",
      gap: 10
    } : t1;
    $[0] = t1;
    $[1] = t2;
  } else {
    t2 = $[1];
  }
  const anchor = t2;
  let t3;
  if ($[2] !== className) {
    t3 = (t42) => {
      const {
        hover,
        open,
        focus
      } = t42;
      return cn("x:cursor-pointer", "x:h-7 x:rounded-md x:px-2 x:text-xs x:font-medium x:transition-colors", open ? "x:bg-gray-200 x:text-gray-900 x:dark:bg-primary-100/10 x:dark:text-gray-50" : hover ? "x:bg-gray-200 x:text-gray-900 x:dark:bg-primary-100/5 x:dark:text-gray-50" : "x:text-gray-600 x:dark:text-gray-400", focus && "x:nextra-focus", className);
    };
    $[2] = className;
    $[3] = t3;
  } else {
    t3 = $[3];
  }
  let t4;
  if ($[4] !== selectedOption || $[5] !== t3 || $[6] !== title) {
    t4 = /* @__PURE__ */ jsx(ListboxButton, { title, className: t3, children: selectedOption });
    $[4] = selectedOption;
    $[5] = t3;
    $[6] = title;
    $[7] = t4;
  } else {
    t4 = $[7];
  }
  let t5;
  if ($[8] !== options) {
    t5 = options.map(_temp2);
    $[8] = options;
    $[9] = t5;
  } else {
    t5 = $[9];
  }
  let t6;
  if ($[10] !== anchor || $[11] !== t5) {
    t6 = /* @__PURE__ */ jsx(ListboxOptions, { as: "ul", transition: true, anchor, className: _temp, children: t5 });
    $[10] = anchor;
    $[11] = t5;
    $[12] = t6;
  } else {
    t6 = $[12];
  }
  let t7;
  if ($[13] !== onChange || $[14] !== t4 || $[15] !== t6 || $[16] !== value) {
    t7 = /* @__PURE__ */ jsxs(Listbox, { value, onChange, children: [
      t4,
      t6
    ] });
    $[13] = onChange;
    $[14] = t4;
    $[15] = t6;
    $[16] = value;
    $[17] = t7;
  } else {
    t7 = $[17];
  }
  return t7;
};
function _temp(t0) {
  const {
    open: open_0
  } = t0;
  return cn("x:focus-visible:nextra-focus", open_0 ? "x:opacity-100" : "x:opacity-0", "x:motion-reduce:transition-none x:transition-opacity x:min-w-(--button-width) x:z-30 x:max-h-64 x:rounded-md x:border x:border-black/5 x:backdrop-blur-md x:bg-nextra-bg/70 x:py-1 x:text-sm x:shadow-lg x:dark:border-white/20");
}
function _temp2(option) {
  return /* @__PURE__ */ jsx(ListboxOption, { value: option.id, as: Fragment, children: (t0) => {
    const {
      selected,
      focus: focus_0
    } = t0;
    return /* @__PURE__ */ jsxs("li", { className: cn(focus_0 ? "x:bg-primary-100 x:text-primary-800 x:dark:text-primary-500 x:dark:bg-primary-500/10" : "x:text-gray-800 x:dark:text-gray-100", "x:cursor-pointer x:whitespace-nowrap x:py-1.5 x:px-3", "x:transition-colors", selected && "x:flex x:items-center x:justify-between x:gap-3"), children: [
      option.name,
      selected && /* @__PURE__ */ jsx(CheckIcon, { height: "1em" })
    ] });
  } }, option.id);
}
export {
  Select
};
