// nextra/dist/server/tsdoc/tsdoc.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import cn from "clsx";
import Slugger from "github-slugger";
import { Callout } from "../../client/components/callout.js";
import { Tabs } from "../../client/components/tabs/index.js";
import { Anchor } from "../../client/mdx-components/anchor.js";
import { Code } from "../../client/mdx-components/code.js";
import { MDXRemote } from "../../client/mdx-remote.js";
import { compileMdx } from "../compile.js";
const Link = ({ className, ...props }) => {
  return /* @__PURE__ */ jsx(
    Anchor,
    {
      className: cn(
        "x:text-primary-600 x:underline x:hover:no-underline x:decoration-from-font x:[text-underline-position:from-font]",
        className
      ),
      ...props
    }
  );
};
async function renderMarkdownDefault(description) {
  if (!description) return;
  const rawJs = await compileMdx(description);
  return /* @__PURE__ */ jsx(MDXRemote, { compiledSource: rawJs });
}
const classes = {
  card: cn(
    "x:rounded-xl nextra-border x:hover:bg-primary-50 x:dark:hover:bg-primary-500/10"
  ),
  anchor: cn(
    "x:absolute x:top-0 x:right-0 x:text-lg x:font-black",
    'x:before:content-["#"] x:hover:text-black x:dark:hover:text-white',
    "x:px-3 x:py-[min(1%,12px)]"
    // Increase click box
  )
};
const TSDoc = ({
  definition,
  renderMarkdown = renderMarkdownDefault,
  typeLinkMap = {},
  noParametersContent = /* @__PURE__ */ jsx(Callout, { type: "info", children: "This function does not accept any parameters." })
}) => {
  if ("entries" in definition) {
    return /* @__PURE__ */ jsx(
      FieldsTable,
      {
        fields: definition.entries,
        typeLinkMap,
        renderMarkdown
      }
    );
  }
  const { signatures, tags } = definition;
  const withSignatures = signatures.length > 1;
  if (!withSignatures) {
    return /* @__PURE__ */ jsx(FunctionSignature, { signature: signatures[0] });
  }
  return /* @__PURE__ */ jsx(
    Tabs,
    {
      items: signatures.map((_, index) => `Function Signature ${index + 1}`),
      children: signatures.map((signature, index) => /* @__PURE__ */ jsx(Tabs.Tab, { children: /* @__PURE__ */ jsx(FunctionSignature, { signature, index: index + 1 }) }, index))
    }
  );
  async function FunctionSignature({
    signature,
    index = ""
  }) {
    const slugger = new Slugger();
    const description = await renderMarkdown(tags?.returns);
    const unnamedReturnId = `returns${index}`;
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx("b", { className: "x:mt-[1.25em] x:block", children: "Parameters:" }),
      signature.params.length ? /* @__PURE__ */ jsx(
        FieldsTable,
        {
          fields: signature.params,
          typeLinkMap,
          renderMarkdown
        }
      ) : noParametersContent,
      /* @__PURE__ */ jsx("b", { className: "x:mt-[1.25em] x:block", children: "Returns:" }),
      Array.isArray(signature.returns) ? /* @__PURE__ */ jsxs(Fragment, { children: [
        description,
        /* @__PURE__ */ jsxs("table", { className: "x:my-6 x:w-full x:text-sm", children: [
          /* @__PURE__ */ jsx("thead", { className: "nextra-border x:border-b x:text-left x:max-lg:hidden", children: /* @__PURE__ */ jsxs("tr", { children: [
            /* @__PURE__ */ jsx("th", { className: "x:py-1.5", children: "Name" }),
            /* @__PURE__ */ jsx("th", { className: "x:p-1.5 x:px-3", children: "Type" })
          ] }) }),
          /* @__PURE__ */ jsx("tbody", { children: signature.returns.map(async (prop) => {
            const id = slugger.slug(prop.name);
            const description2 = await renderMarkdown(
              prop.description || prop.tags?.description
            );
            return /* @__PURE__ */ jsxs(Row, { id, children: [
              /* @__PURE__ */ jsx(
                NameCell,
                {
                  id,
                  optional: prop.optional,
                  name: prop.name
                }
              ),
              /* @__PURE__ */ jsx(
                TypeAndDescriptionCell,
                {
                  type: prop.type,
                  description: description2,
                  typeLinkMap
                }
              )
            ] }, id);
          }) })
        ] })
      ] }) : /* @__PURE__ */ jsxs(
        "div",
        {
          id: unnamedReturnId,
          className: cn(
            classes.card,
            'x:text-sm x:relative x:p-3 x:border x:before:content-["Type:_"] x:mt-5'
          ),
          children: [
            /* @__PURE__ */ jsx("a", { href: `#${unnamedReturnId}`, className: cn(classes.anchor) }),
            /* @__PURE__ */ jsx(Code, { children: linkify(signature.returns.type, typeLinkMap) }),
            description && /* @__PURE__ */ jsx("div", { className: "x:mt-2", children: description })
          ]
        }
      )
    ] });
  }
};
const Row = ({ children, id }) => {
  return /* @__PURE__ */ jsx(
    "tr",
    {
      id,
      className: cn(
        classes.card,
        "x:group x:mb-5 x:max-lg:block x:max-lg:border x:lg:border-b",
        "x:lg:not-target:[&>td>a]:opacity-0"
      ),
      children
    }
  );
};
const NameCell = ({ name, id, optional }) => {
  return /* @__PURE__ */ jsxs(
    "td",
    {
      className: cn(
        "x:relative x:max-lg:block",
        name && "x:py-3 x:max-lg:px-3"
      ),
      children: [
        /* @__PURE__ */ jsx(
          "a",
          {
            href: `#${id}`,
            className: cn(
              classes.anchor,
              "x:group-hover:opacity-100! x:lg:top-1/2 x:lg:right-full x:lg:-translate-y-1/2"
            )
          }
        ),
        name && /* @__PURE__ */ jsx(
          Code,
          {
            className: cn(
              "x:max-md:break-all",
              // add `?` via CSS `content` property so value will be not selectable
              optional && 'x:after:content-["?"]'
            ),
            children: name
          }
        )
      ]
    }
  );
};
const TypeAndDescriptionCell = ({ type, description, typeLinkMap }) => {
  return /* @__PURE__ */ jsxs(
    "td",
    {
      className: 'x:p-3 x:max-lg:block x:max-lg:before:content-["Type:_"]',
      children: [
        /* @__PURE__ */ jsx(Code, { children: linkify(type, typeLinkMap) }),
        description && /* @__PURE__ */ jsx("div", { className: "x:mt-2", children: description })
      ]
    }
  );
};
const FieldsTable = ({ fields, typeLinkMap, renderMarkdown }) => {
  const slugger = new Slugger();
  return /* @__PURE__ */ jsxs("table", { className: "x:my-8 x:w-full x:text-sm", children: [
    /* @__PURE__ */ jsx("thead", { className: "nextra-border x:border-b x:text-left x:max-lg:hidden", children: /* @__PURE__ */ jsxs("tr", { children: [
      /* @__PURE__ */ jsx("th", { className: "x:py-1.5", children: "Name" }),
      /* @__PURE__ */ jsx("th", { className: "x:p-1.5 x:px-3", children: "Type" }),
      /* @__PURE__ */ jsx("th", { className: "x:py-1.5", children: "Default" })
    ] }) }),
    /* @__PURE__ */ jsx("tbody", { children: fields.map(async (field) => {
      const id = slugger.slug(field.name);
      const tags = field.tags ?? {};
      const defaultValue = tags.default || tags.defaultValue;
      const description = await renderMarkdown(
        [
          field.description || tags.description,
          tags.deprecated && `**Deprecated**: ${tags.deprecated}`
        ].filter(Boolean).join("\n")
      );
      return /* @__PURE__ */ jsxs(Row, { id, children: [
        /* @__PURE__ */ jsx(NameCell, { id, optional: field.optional, name: field.name }),
        /* @__PURE__ */ jsx(
          TypeAndDescriptionCell,
          {
            type: field.type,
            description,
            typeLinkMap
          }
        ),
        /* @__PURE__ */ jsx(
          "td",
          {
            className: cn(
              "x:max-lg:block",
              // For the mobile view, we want to hide the default column entirely if there is no
              // content for it. We want this because otherwise, the default padding applied to
              // table cells will add some extra blank space we don't want.
              defaultValue ? (
                // add `Default: ` via CSS `content` property so value will be not selectable
                'x:py-3 x:max-lg:pt-0 x:max-lg:px-3 x:max-lg:before:content-["Default:_"]'
              ) : 'x:lg:after:content-["\u2013"]'
            ),
            children: defaultValue && /* @__PURE__ */ jsx(Code, { className: "x:whitespace-pre-wrap x:inline-block", children: linkify(defaultValue, typeLinkMap) })
          }
        )
      ] }, id);
    }) })
  ] });
};
function linkify(type, typeLinkMap = {}) {
  const result = [];
  for (const chunk of type.match(/(\w+|\W+)/g)) {
    const href = typeLinkMap[chunk];
    if (href) {
      result.push(
        /* @__PURE__ */ jsx(Link, { href, children: /* @__PURE__ */ jsx(Fragment, { children: chunk }) }, result.length)
      );
      continue;
    }
    if (typeof result.at(-1) === "string") {
      result[result.length - 1] += chunk;
      continue;
    }
    result.push(chunk);
  }
  return result;
}
export {
  TSDoc
};
